File: custom_dir/custom_actions/random_task.py
# -*- coding: UTF-8 -*-
import json
import random

from maa.context import Context
from maa.custom_action import CustomAction
from maa.agent.agent_server import AgentServer


@AgentServer.custom_action("RandomTask")
class RandomTask(CustomAction):
    def run(self,
            context: Context,
            argv: CustomAction.RunArg) -> bool:
        """
        自定义动作：随机任务
        :param argv: custom_action_param 格式：{"任务1":100,"任务2":300,"任务3":600}
        :param context: 执行上下文
        :return: 是否执行成功
        """

        print("开始执行自定义动作：随机任务")

        try:
            json_data = json.loads(argv.custom_action_param)
        except Exception as e:
            print(f"参数解析失败: {e}")
            return False

        if not json_data:
            print("传入参数为空")
            return False

        # 分离任务和权重
        tasks = list(json_data.keys())
        weights = list(json_data.values())

        # 随机选择一个任务
        chosen_task = random.choices(tasks, weights=weights, k=1)[0]
        print(f"随机选择的任务是: {chosen_task}")

        # 执行任务
        try:
            context.run_task(chosen_task)
            return True
        except Exception as e:
            print(f"执行任务失败: {e}")
            return False

    def stop(self) -> None:
        """停止逻辑"""
        pass

File: agent/actions/random_task/random_task.go
package random_task

import (
	"encoding/json"
	"fmt"
	"math/rand"

	"github.com/MaaXYZ/maa-framework-go/v4"
)

type RandomTask struct{}

// Run 根据权重随机选择并执行一个任务
func (a *RandomTask) Run(ctx *maa.Context, arg *maa.CustomActionArg) bool {
	fmt.Println("开始执行自定义动作：随机任务")

	// 解析参数
	var taskWeightsMap map[string]float64
	if err := json.Unmarshal([]byte(arg.CustomActionParam), &taskWeightsMap); err != nil {
		fmt.Printf("参数JSON解析失败: %v\n", err)
		return false
	}

	if len(taskWeightsMap) == 0 {
		fmt.Println("传入的任务列表为空")
		return false
	}

	// 分离任务和权重
	tasks := make([]string, 0, len(taskWeightsMap))
	weights := make([]float64, 0, len(taskWeightsMap))
	for task, weight := range taskWeightsMap {
		if weight < 0 {
			fmt.Printf("错误：权重值不能为负数。发现无效权重: %f\n", weight)
			return false
		}
		tasks = append(tasks, task)
		weights = append(weights, weight)
	}

	// 校验权重总和
	var totalWeight float64
	for _, w := range weights {
		totalWeight += w
	}
	if totalWeight <= 0 {
		fmt.Println("错误：所有任务的权重总和必须大于0。")
		return false
	}

	// 随机选择任务
	fmt.Printf("任务池: %v\n", tasks)
	fmt.Printf("对应权重: %v\n", weights)

	chosenIdx := weightedRandomChoice(weights)
	chosenTask := tasks[chosenIdx]
	fmt.Printf("随机选择的任务是: %s\n", chosenTask)

	// 执行任务
	_, err := ctx.RunTask(chosenTask)
	if err != nil {
		fmt.Printf("任务 '%s' 执行失败: %v\n", chosenTask, err)
		return false
	}
	fmt.Printf("任务 '%s' 执行成功。\n", chosenTask)

	return true
}

// weightedRandomChoice 根据权重随机选择索引
func weightedRandomChoice(weights []float64) int {
	var totalWeight float64
	for _, w := range weights {
		totalWeight += w
	}

	r := rand.Float64() * totalWeight
	var sum float64
	for i, w := range weights {
		sum += w
		if r <= sum {
			return i
		}
	}
	return len(weights) - 1
}